Fix to list paste bug

Quintis1212 opened the pull request:
Hello ) this is a fix to "Gutenberg incorrect paste of list text into list block #26713" issue , I added ul and div tags to clipboardData for correct paste of lists

Quintis1212 committed:
fix to list paste bug

Quintis1212 committed:
fix regexp

grzim commented on packages/rich-text/src/component/index.js:
```diff
@@ -325,8 +325,32 @@ function RichText(
+				.join( '' )
+				.split( '</li>' )
+				.filter( Boolean );
+			const liCopyStringLength = liCopyString.length - 1;
```
there is no need of adding this variable here, we can get a length of it inside `map` operator

```js
.map( ( el, i, {length} ) => {
  if ( el && i === 0 ) {
	  return '<ul><li>' + el + '</li>';
  } else if ( el && i === length - 1 ) {
	  return '<div style="clear:both"/><li>' + el + '</li></ul>';
  }
  return '<div style="clear:both"/><li>' + el + '</li>';
} );
```

Quintis1212 replied:
nice tip ) can you give me a link for tips like this , I first time see that I can get length in arguments of callback) where did you get this tip ?)

grzim replied:
uummm... this is just an object destructuring (an array is an object, and it is passed as the third parameter to the callback). 
Please check the documentation https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map

grzim commented on packages/rich-text/src/component/index.js:
```diff
@@ -325,8 +325,32 @@ function RichText(
+				if ( el && i === 0 ) {
+					return '<ul><li>' + el + '</li>';
+				} else if ( el && i === liCopyStringLength ) {
+					return '<div style="clear:both"/><li>' + el + '</li></ul>';
```
The direct children of `<ul>` should be only `<li>` and not `<div>`. The generated HTML would be semantically incorrect. 

Please check https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul

Quintis1212 replied:
according to yours link I can paste template tag , I need something between li tags to push them to new lines

grzim replied:
cant you just style li elements? Btw, it is better not to add inline styles, but CSS classes.

Quintis1212 replied:
> cant you just style li elements? Btw, it is better not to add inline styles, but CSS classes.

I ve tested this approach , it must be something between li elements .There is no css or scss file in components folder , where I must write down css for class ?

grzim commented on packages/rich-text/src/component/index.js:
```diff
@@ -325,8 +325,32 @@ function RichText(
+				} else if ( el && i === liCopyStringLength ) {
+					return '<div style="clear:both"/><li>' + el + '</li></ul>';
+				}
+				return '<div style="clear:both"/><li>' + el + '</li>';
```
if `el` could be falsy (in two `if` statements above we check for it) and would be for the first element, then it will fall to the last case `'<div style="clear:both"/><li>' + el + '</li>'` what would generate improper HTML ( `<ul>` will be missing)

Quintis1212 replied:
I added a check for last return in my commit , but my lint checker log me an error :
"
  337:59  error  Array.prototype.map() expects a value to be returned at the end of arrow function  array-callback-return

"
It seems it must be some return in the end of function , or I can disable error with comment

grzim replied:
If it is possible that `el` is empty why not filter them out before mapping?

Quintis1212 replied:
i ve filter all elements in .filter( Boolean ) function before mapping

grzim commented on packages/rich-text/src/component/index.js:
```diff
@@ -325,8 +325,32 @@ function RichText(
-		event.clipboardData.setData( 'text/html', html );
+
+		const isLi = /<li>.*<\/li>/gi;
+		const foundLi = html.match( isLi );
```
it is better to use a boolean value here with proper semantic:
`const hasLi = html.match( isLi )?.length > 0;`

Quintis1212 replied:
I changed my code with your comment , can you give me please a link for documentation of use this question mark after ( isLi ). Is this lodash library ?

grzim replied:
It is called optional chaining and it is a native js feature https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

grzim commented:
Thank @Quintis1212 for your effort. I have added a few comments to your code 👍

Quintis1212 committed:
improved version of fix

Quintis1212 commented:
> Thank @Quintis1212 for your effort. I have added a few comments to your code

@grzim   Thank you for help !) It seems there is new errors after I changed my code according to yours comments : 


`
[1] /home/runner/work/gutenberg/gutenberg/packages/rich-text/src/component/index.js
[1]   334:47  error  Replace `length` with `·length·`                                                                                                                                                   prettier/prettier
[1]   334:57  error  Array.prototype.map() expects a value to be returned at the end of arrow function                                                                                                  array-callback-return
[1]   338:13  error  Replace `'<template·style="clear:both"></template><li>'·+·el·+·'</li></ul>'` with `(⏎↹↹↹↹↹↹'<template·style="clear:both"></template><li>'·+⏎↹↹↹↹↹↹el·+⏎↹↹↹↹↹↹'</li></ul>'⏎↹↹↹↹↹)`  prettier/prettier
[1]   340:13  error  Replace `'<template·style="clear:both"></template><li>'·+·el·+·'</li>'` with `(⏎↹↹↹↹↹↹'<template·style="clear:both"></template><li>'·+⏎↹↹↹↹↹↹el·+⏎↹↹↹↹↹↹'</li>'⏎↹↹↹↹↹)`            prettier/prettier

`
please , can you help me to fix this errors ?

grzim commented:
You're welcome!
These are linting errors. You can run prettier it will format the code for you. 
It is already a part of a project and you can run it for the desired file https://prettier.io/docs/en/cli.html

Quintis1212 commented:
> You're welcome!
> These are linting errors. You can run prettier it will format the code for you.
> It is already a part of a project and you can run it for the desired file https://prettier.io/docs/en/cli.html

But prettier is already be runned in pre commit hook , I tried to run prettier index.js in terminal but there is no such command in terminal

Mamaduka commented:
I'm going to close this PR. With the new list block, we'll need an alternative fix for this bug.

Thank you for your contribution, @Quintis1212!
